data_model/src/trigger.rs

```diff
@@ -2,43 +2,154 @@
 )]
-pub struct Trigger {
+pub struct Trigger<F: Filter> {
     /// [`Id`] of the [`Trigger`].
-    pub id: <Self as Identifiable>::Id,
+    pub id: Id,
     /// Action to be performed when the trigger matches.
-    pub action: Action,
+    pub action: Action<F>,
 }
 
-impl Trigger {
+impl<F: Filter> Trigger<F> {
     /// Construct trigger, given name action and signatories.
-    pub fn new(
-        id: <Self as Identifiable>::Id,
-        action: Action,
-    ) -> <Self as Identifiable>::RegisteredWith {
+    pub fn new(id: Id, action: Action<F>) -> Self {
         Self { id, action }
     }
 }
 
-impl Identifiable for Trigger {
+impl TryFrom<Trigger<FilterBox>> for Trigger<DataEventFilter> {
+    type Error = &'static str;
+
+    fn try_from(boxed: Trigger<FilterBox>) -> Result<Self, Self::Error> {
+        if let FilterBox::Data(data_filter) = boxed.action.filter {
+            let action = Action::new(
+                boxed.action.executable,
+                boxed.action.repeats,
+                boxed.action.technical_account,
+                data_filter,
+            );
+            Ok(Self {
+                id: boxed.id,
+                action,
+            })
+        } else {
+            Err("Expected `FilterBox::Data`, but another variant found")
+        }
+    }
+}
+
+impl TryFrom<Trigger<FilterBox>> for Trigger<PipelineEventFilter> {
+    type Error = &'static str;
+
+    fn try_from(boxed: Trigger<FilterBox>) -> Result<Self, Self::Error> {
+        if let FilterBox::Pipeline(pipeline_filter) = boxed.action.filter {
+            let action = Action::new(
+                boxed.action.executable,
+                boxed.action.repeats,
+                boxed.action.technical_account,
+                pipeline_filter,
+            );
+            Ok(Self {
+                id: boxed.id,
+                action,
+            })
+        } else {
+            Err("Expected `FilterBox::Pipeline`, but another variant found")
+        }
+    }
+}
+
+impl TryFrom<Trigger<FilterBox>> for Trigger<TimeEventFilter> {
+    type Error = &'static str;
+
+    fn try_from(boxed: Trigger<FilterBox>) -> Result<Self, Self::Error> {
+        if let FilterBox::Time(time_filter) = boxed.action.filter {
+            let action = Action::new(
+                boxed.action.executable,
+                boxed.action.repeats,
+                boxed.action.technical_account,
+                time_filter,
+            );
+            Ok(Self {
+                id: boxed.id,
+                action,
+            })
+        } else {
+            Err("Expected `FilterBox::Time`, but another variant found")
+        }
+    }
+}
+
+impl TryFrom<Trigger<FilterBox>> for Trigger<ExecuteTriggerEventFilter> {
+    type Error = &'static str;
+
+    fn try_from(boxed: Trigger<FilterBox>) -> Result<Self, Self::Error> {
+        if let FilterBox::ExecuteTrigger(execute_trigger_filter) = boxed.action.filter {
+            let action = Action::new(
```

Comment:
```suggestion
            let action = Action {
            	..boxed.action
            }
```
Could be idiomatically what you're looking for.

Reply:
Not working cause of different action types :(